simple_network_sim/inference.py

```diff
@@ -0,0 +1,708 @@
+    sampled and eventually inferred. To be used within ABC-SMC, a particle
+    must satisfy the following requirements:
+    1) We can sample from the priors and get the prior pdf
+    2) We should be able to generate a perturbated particle
+    """
+
+    inferred_variables_classes: ClassVar[Dict[str, Type[InferredVariable]]] = {
+        "infection-probability": InferredInfectionProbability,
+        "initial-infections": InferredInitialInfections
+    }
+
+    def __init__(self, inferred_variables: Dict[str, InferredVariable]):
+        self.inferred_variables = inferred_variables
+
+    @staticmethod
+    def generate_from_priors(fitter: ABCSMC) -> Particle:
+        """ Generate a particle from the prior distribution. All parameters in the
+        particle are independent and therefore we simply random each parameter from
+        its prior distribution. The fitter object provides the random state used for
+        random variable generation but also the prior parameters, which are
+        the values in the data pipeline.
+
+        :return: New particle generated from prior
+        """
+        return Particle({variable_name: variable_class.generate_from_prior(fitter)
+                         for variable_name, variable_class in Particle.inferred_variables_classes.items()})
+
+    def generate_perturbated(self) -> Particle:
+        """ From current particle, add a perturbation to every parameter and return
+        a newly created perturbated particle.
+
+        :return: New particle which is similar to self up to a perturbation
+        """
+        perturbed_variables = {}
+        for name, inferred_variable in self.inferred_variables.items():
+            perturbed_variables[name] = inferred_variable.generate_perturbated()
+
+        return Particle(perturbed_variables)
+
+    def validate_particle(self) -> bool:
+        """ Checks that the particle is valid, simply checks that all parameters inside
+        respect their supports, i.e. the values are attainable. It may return False for
+        example if the uniform perturbation pushed a lognormally-distribution value
+        below 0.
+
+        :return: Whether the particle is valid
+        """
+        return all(variable.validate() for variable in self.inferred_variables.values())
+
+    @staticmethod
+    def resample_and_perturbate(
+            particles: List[Particle],
+            weights: List[float],
+            random_state: np.random.Generator
+    ) -> Particle:
+        """ Resampling part of ABC-SMC, selects randomly a new particle from the
+        list of previously accepted. Then perturb it slightly. This causes
+        the persistence and selection of fit particles in a manner similar to
+        evolution algorithms. The perturbation is done until the candidate is valid,
+        this is because the perturbation is unaware of the support of the distribution
+        of the particle.
```

Comment:
This is fine for us here, and might be unavoidable, but I can imagine circumstances in which this causes a lot of extra perturbation.  Nothing I suggest you change, just a note.

Reply:
Yes I had to fight a bit against some of these circumstances. For example when parameters are close the the edge of their supports many perturbations will push them beyond, which will require a resampling. Thankfully each of these is bunch of numpy calls and do not require model evaluation, so they don't consume much time. This problem can be mitigated with being careful with the priors.

E.g. the problem I had was when I used a lognormal prior for initial infections, and they were at 0 for some HBs (0 is not in the support for the lognormal distributions). So I bumped the priors from 0 to 1 with a small std dev.